### ISSUE TITLE
[BUG] Multi-GPU performance worse than single GPU

### ISSUE THREAD

**Issue by pratogab** [microsoft/DeepSpeed#4550]
I will refer to the [issue](https://github.com/huggingface/accelerate/issues/2027) I opened on the accelerate github, because I am unsure if this is an accelerate, Deep Speed or torch issue. All of the details are there.

**Comment by GabPrato** [comment]
**TL;DR:** The performance when training a model with DeepSpeed is slightly worse in multi-GPU setup compared to single GPU, depending on the optimizer, and the reason seems to be because DeepSpeed does not share the moving averages between devices, for example the first and second moment estimates of Adam. This is true whether or not you use torch or DeepSpeed's optimizers.

After looking further into this issue, I think I have figured out what is going on. While I don't have a definite proof, I am sure someone more knowledgeable with the DeepSpeed library can confirm if this is correct or not.

Here's what I observed, running accelerate with DeepSpeed, 1 vs 2 GPUs
- First training step, everything is the same, model input, model weights, model output, loss, gradient, model weights after the weight update.
- Second training step, model input, model weights, model output, loss and gradient are the same. But then, the model weights after the weight update differ between the 1 vs 2 GPU setup.
- Then, for the following steps, since the weights have diverged, everything else diverges and the performance ends up worse for the 2 GPU setup compared to the 1 GPU setup.

I believe that this is likely because of the issue I have mentioned in the TL;DR: DeepSpeed probably does not share the moving averages of optimizers between devices, which leads to the divergence I am observing.

Here's how one can reproduce this. I am using huggingface accelerate, but surely using solely DeepSpeed would result in the same problem.

OS and devices:
A100 GPUs, Ubuntu 22.04, CUDA 11.8

Packages:
```
accelerate=0.25.0
deepspeed=0.12.6
python=3.10.13
pytorch=2.1.2
```

Code to reproduce:
```python
import argparse
import torch
import torch.nn as nn
from accelerate import Accelerator
from accelerate.utils import set_seed, DummyOptim


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.w = nn.Linear(1, 1)

    def forward(self, x):
        return self.w(x)


class Dataset(torch.utils.data.IterableDataset):
    def __iter__(self):
        for _ in range(1000000000):
            x = torch.rand(1) * 2
            y = torch.zeros(1) if x < 1 else torch.ones(1)
            yield x, y

    def __len__(self):
        return 1000000000


class DataLoader(torch.utils.data.DataLoader):
    def __init__(self, batch_size):
        super().__init__(Dataset(), batch_size=batch_size, shuffle=False)

    def __iter__(self):
        for batch in self.dataset:
            yield batch


parser = argparse.ArgumentParser()
parser.add_argument('--batch_size', type=int)
parser.add_argument('--steps', type=int, default=10)
parser.add_argument('--seed', type=int, help='If not None, will set the seed for random, numpy, torch, torch.cuda and if TPUs are available torch_xla’s cuda state.')
args = parser.parse_args()

accelerator = Accelerator(
    split_batches=True,
    dispatch_batches=True)

if args.seed is not None:
    set_seed(args.seed, device_specific=True)

model = Model().to(accelerator.device)
optimizer = DummyOptim(model.parameters(), lr=1e-1, weight_decay=0)
criterion = nn.BCEWithLogitsLoss()
dataloader = DataLoader(batch_size=args.batch_size)

model, optimizer, dataloader = accelerator.prepare(model, optimizer, dataloader)

for step, batch in enumerate(dataloader):
    batch, target = batch
    batch = batch.to(dtype=model.w.weight.dtype)
    print(accelerator.local_process_index, batch, target)

    print(accelerator.local_process_index, model.w.weight.data, model.w.bias.data)

    output = model(batch)
    print(accelerator.local_process_index, output)
    loss = criterion(output, target)
    print(accelerator.local_process_index, loss.item())
    accelerator.backward(loss)
    print(accelerator.local_process_index, model.w.weight.grad, model.w.bias.grad)
    print(accelerator.local_process_index, model.w.weight.data, model.w.bias.data)
    optimizer.step()
    optimizer.zero_grad()

    if (step + 1) == args.steps:
        break
```
Feel free to comment out `print` calls. Printing the weights after each update is enough to see the divergence.

Here's my accelerate config for the 1 GPU case:
```yaml
compute_environment: LOCAL_MACHINE
debug: false
deepspeed_config:
  deepspeed_config_file: deespeed_config.json
  zero3_init_flag: false
distributed_type: DEEPSPEED
downcast_bf16: 'no'
machine_rank: 0
main_training_function: main
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```
And its deepspeed config:
```json
{
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": 1e-1,
            "betas": [
                0.9,
                0.999
            ],
            "eps": 1e-8,
            "weight_decay": 0
        }
    },
    "zero_optimization": {
        "stage": 0,
        "offload_optimizer": {
            "device": "none",
            "nvme_path": null
        },
        "offload_param": {
            "device": "none",
            "nvme_path": null
        }
    },
    "bf16": {
        "enabled": true
    },
    "fp16": {
        "enabled": false
    },
    "train_batch_size": 64,
    "train_micro_batch_size_per_gpu": 64,
    "gradient_accumulation_steps": 1
}
```

And the 2 gpu accelerate config:
```yaml
compute_environment: LOCAL_MACHINE
debug: false
deepspeed_config:
  deepspeed_config_file: deespeed_config_B.json
  zero3_init_flag: false
distributed_type: DEEPSPEED
downcast_bf16: 'no'
machine_rank: 0
main_training_function: main
num_machines: 1
num_processes: 2
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

And its deepspeed config:
```json
{
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": 1e-1,
            "betas": [
                0.9,
                0.999
            ],
            "eps": 1e-8,
            "weight_decay": 0
        }
    },
    "zero_optimization": {
        "stage": 0,
        "offload_optimizer": {
            "device": "none",
            "nvme_path": null
        },
        "offload_param": {
            "device": "none",
            "nvme_path": null
        }
    },
    "bf16": {
        "enabled": true
    },
    "fp16": {
        "enabled": false
    },
    "train_batch_size": 64,
    "train_micro_batch_size_per_gpu": 32,
    "gradient_accumulation_steps": 1
}
```

Now if you run the python script for both the 1 gpu and 2 gpu case:
`accelerate launch --config_file=<accelerate_config_1gpu> <python_script> --batch_size=64 --steps=500 --seed=0`
`accelerate launch --config_file=<accelerate_config_2gpus> <python_script> --batch_size=64 --steps=500 --seed=0`
You can compare the model weights and observe what I have described.

I ran seeds 0 to 4 included, averaged the loss and get the following graph:
![temp_deepspeed_example](https://github.com/microsoft/DeepSpeed/assets/25964820/d9e6b288-a83f-4bb3-a8a3-e403af89019d)
This showcases the performance difference, where the orange loss is doing slightly worse than the blue loss. Note that if I change the optimizer to SGD, then the lines are identical, since SGD doesn't have moving averages.

Only tested with Adam, but I assume that any optimizer with moving averages has this problem. Maybe this is a known problem for the DeepSpeed developers and potentially a design choice. It might be too slow to share moving averages. Maybe it should be an option for the user.

I might also be wrong with respect to the cause of the problem, but someone more knowledgeable than me with DeepSpeed can surely confirm whether this is the problem or not.

If you need any other details, let me know.

**Comment by GabPrato** [comment]
FYI, I closed the issue that was opened on accelerate's side because I had two issues opened, one in accelerate's repo and one in DeepSpeed's repo, because I didn't know at the time where the problem was.

**Comment by olegsinavski** [comment]
Hello, I observe the same except way worse - it diverges when I bump a number of GPUs. I'm training BLIP-2 model from scratch in bf16 precision.

<img width="709" alt="image" src="https://github.com/microsoft/DeepSpeed/assets/2086260/5b86d97f-80b1-4713-8621-f24558c2ed4a">

**Comment by Martin7-1** [comment]
Any update on this issue? I observe the same thing.

**Comment by whcjb** [comment]
any update?

**Comment by tjruwase** [comment]
@GabPrato, apologies for the delayed response. I am unable to reproduce the divergence on 1GPU vs 2GPU. I followed the stesp you shared and compared experiments using seed values 0, 1, 2, 3, 4, as suggested. Below is the tensorboard of loss values across the ten runs. As you can see the final loss values are aligned for each seed setting. 

<img width="1037" alt="Image" src="https://github.com/user-attachments/assets/fb50c012-a037-4e5c-ad77-90908bd0a33b" />


A few more details on my run.
1. Package versions
```bash
accelerate               0.34.0
deepspeed                0.16.2
transformers             4.47.0
```
2. I modified your code with the key change being to compute average loss as `torch.mean(accelerator.gather(loss))`. Full code below.

```python
import argparse
import torch
import torch.nn as nn
from accelerate import Accelerator
from accelerate.utils import set_seed, DummyOptim
import os
from torch.utils.tensorboard import SummaryWriter

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.w = nn.Linear(1, 1)

    def forward(self, x):
        return self.w(x)


class Dataset(torch.utils.data.IterableDataset):
    def __iter__(self):
        for _ in range(1000000000):
            x = torch.rand(1) * 2
            y = torch.zeros(1) if x < 1 else torch.ones(1)
            yield x, y

    def __len__(self):
        return 1000000000


class DataLoader(torch.utils.data.DataLoader):
    def __init__(self, batch_size):
        super().__init__(Dataset(), batch_size=batch_size, shuffle=False)

    def __iter__(self):
        for batch in self.dataset:
            yield batch


parser = argparse.ArgumentParser()
parser.add_argument('--batch_size', type=int)
parser.add_argument('--steps', type=int, default=10)
parser.add_argument('--seed', type=int, help='If not None, will set the seed for random, numpy, torch, torch.cuda and if TPUs are available torch_xla’s cuda state.')
args = parser.parse_args()

accelerator = Accelerator(
    split_batches=True,
    dispatch_batches=True)

if args.seed is not None:
    set_seed(args.seed, device_specific=True)

model = Model().to(accelerator.device)
optimizer = DummyOptim(model.parameters(), lr=1e-1, weight_decay=0)
criterion = nn.BCEWithLogitsLoss()
dataloader = DataLoader(batch_size=args.batch_size)

tb_dir=os.path.join('runs', f'seed{args.seed}_{accelerator.num_processes}gpus')
writer = SummaryWriter(log_dir=tb_dir)

model, optimizer, dataloader = accelerator.prepare(model, optimizer, dataloader)

for step, batch in enumerate(dataloader):
    batch, target = batch
    batch = batch.to(dtype=model.w.weight.dtype)
    # print(accelerator.local_process_index, batch, target)

    # print(accelerator.local_process_index, model.w.weight.data, model.w.bias.data)

    output = model(batch)
    # print(f"output = {accelerator.local_process_index}, {output}")
    loss = criterion(output, target)
    all_losses = accelerator.gather(loss)
    # print(f"loss = {accelerator.local_process_index}, {loss.item()}")
    if accelerator.local_process_index == 0:
        print(f"averaged loss [{accelerator.local_process_index}, {step}] =  {torch.mean(all_losses)}")
        writer.add_scalar('Loss/train', torch.mean(all_losses), step)
        
    accelerator.backward(loss)
    # print(accelerator.local_process_index, model.w.weight.grad, model.w.bias.grad)
    # print(accelerator.local_process_index, model.w.weight.data, model.w.bias.data)
    optimizer.step()
    optimizer.zero_grad()

    if (step + 1) == args.steps:
        break
```

**Comment by tjruwase** [comment]
> Hello, I observe the same except way worse - it diverges when I bump a number of GPUs. I'm training BLIP-2 model from scratch

@olegsinavski, since you running a different model, can you please create a new ticket with repro instructions? Thanks!

**Comment by tjruwase** [comment]
@Martin7-1 and @whcjb, since I am unable to repro using the ticket details, can you please share your repro steps? Thanks!

**Comment by tjruwase** [comment]
@GabPrato, I appreciate your response despite the issue being old. Apologies for not responding earlier. I will close the ticket for now. Feel free to re-open if your testing finds something different. I have requested other users to open new tickets since their workload seems to be different.
